llvm/tools/repo2obj/repo2obj.cpp

```diff
@@ -539,23 +594,34 @@ int main(int argc, char *argv[]) {
 
   State.writeGroupSections(OS);
 
-  // Write the string table (and patch its section header)
-  {
-    auto &S = State.SectionHeaders[SectionIndices::StringTab];
-    std::tie(S.sh_offset, S.sh_size) = State.Strings.write(OS);
-  }
   // Now do the same for the symbol table.
   {
     auto &S = State.SectionHeaders[SectionIndices::SymTab];
     // st_info should be one greater than the symbol table index of the last
     // local symbol (binding STB_LOCAL).
     S.sh_info = SymbolTable<ELFT>::firstNonLocal(OrderedSymbols);
-    std::tie(S.sh_offset, S.sh_size) = State.Symbols.write(OS, OrderedSymbols);
+    bool NeedsExtendedShdnx = false;
+    std::tie(S.sh_offset, S.sh_size, NeedsExtendedShdnx) =
+        State.Symbols.write(OS, OrderedSymbols);
+    if (NeedsExtendedShdnx) {
+      // extended symbol table section indices
+      size_t const SymtabShndx = State.addSymtabShndxSection();
+      auto &SymtabShndxSection = State.SectionHeaders[SymtabShndx];
+      std::tie(SymtabShndxSection.sh_offset, SymtabShndxSection.sh_size) =
+          State.Symbols.writeSymtabShndx(OS, OrderedSymbols);
+    }
+  }
+
+  // Write the string table (and patch its section header)
+  {
+    auto &S = State.SectionHeaders[SectionIndices::StringTab];
+    std::tie(S.sh_offset, S.sh_size) = State.Strings.write(OS);
   }
 
-  uint64_t SectionHeadersOffset = State.writeSectionHeaders(OS);
-  Header.e_shoff = SectionHeadersOffset;
-  Header.e_shnum = State.SectionHeaders.size();
+  State.setEShNum(Header);
+  Header.e_shoff = State.writeSectionHeaders(OS);
+  static_assert(SectionIndices::StringTab < llvm::ELF::SHN_LORESERVE,
+                "String table index should be less than LORESERVE");
   Header.e_shstrndx = SectionIndices::StringTab;
```

Comment:
In the document, it is mentioned `ELF Header: e_shstrndx`:

> If` the section name string table index is greater than SHN_LORESERVE (0xff00), this member has the value SHN_XINDEX (0xffff), and the actual index of the section name string table is in the member sh_link of the section header at index 0. (Both entries will be zero if there is no section name string table.) 

Should the `Ehdr.e_shstrndx` look like below?

```
if (the section name string table index  >= SHN_LORESERVE)
    Ehdr.e_shstrndx = SHN_XINDEX;
else
    Ehdr.e_shstrndx = the section name string table index;
```

Reply:
> Should the Ehdr.e_shstrndx look like below?

It’s not necessary because `SectionIndices::StringTab` is a small constant. Line 623 holds a static assertion that the value is < `SHN_LORESERVE`.

Reply:
Thanks, I see.